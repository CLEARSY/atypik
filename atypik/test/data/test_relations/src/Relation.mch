/* Relation
 * Author: vdemeulenaere
 * Creation date: 23/05/2023
 */
MACHINE
    Relation

ABSTRACT_VARIABLES
    var, var2, var3, inv, proj, proj2,
    direct_prod, paral_prod, iter, rfl_closure,
    closure_var, var4, var5, var6, var7, var8
    
INVARIANT
    var : POW(POW(INTEGER * BOOL)) &
    var2 : POW(INTEGER * INTEGER) &
    var3 : POW(BOOL * INTEGER) &
    inv : POW(BOOL * INTEGER) &
    proj : POW(INTEGER * BOOL * INTEGER) &
    proj2 : POW(INTEGER * BOOL * BOOL) &
    direct_prod : POW(BOOL * (INTEGER * REAL)) &
    paral_prod : POW((BOOL * REAL) * (INTEGER * BOOL)) &
    iter : POW(INTEGER * INTEGER) &
    rfl_closure : POW(INTEGER * INTEGER) &
    closure_var : POW(INTEGER * INTEGER) &
    dom(var3) = {TRUE, FALSE} &
    ran(var3) = {1,2,3} &
    inv[{TRUE, FALSE}] = {1, 2} &
    var4 : POW(BOOL * INTEGER) &
    var5 : POW(BOOL * INTEGER) &
    var6 : POW(BOOL * INTEGER) &
    var7 : POW(BOOL * INTEGER) &
    var7 : POW(BOOL * INTEGER) &
    var8 : POW(BOOL * INTEGER)
    

INITIALISATION
    var := 0..5 <-> BOOL ||
    var2 := id({1,2,3}) ||
    inv := {(0 |-> FALSE), (1 |-> TRUE)}~ ||
    proj := prj1({0,1}, {FALSE, TRUE}) || 
    proj2 := prj2({0,1}, {FALSE, TRUE}) ||
    var3 := (inv ; var2) ||
    direct_prod := var3 >< {(FALSE |-> 0.0), (TRUE |-> 1.0)} ||
    paral_prod := (var3 || {(0.0 |-> FALSE), (1.0 |-> TRUE)}) ||
    iter := iterate(var2, 2) ||
    rfl_closure := closure(var2) ||
    closure_var := closure1(var2) ||
    var4 := BOOL <<| inv ||
    var5 := {FALSE, TRUE} <| inv ||
    var6 := var5 |> INTEGER ||
    var7 := var6 |>> {1,2} ||
    var8 := var6 <+ var7

END
