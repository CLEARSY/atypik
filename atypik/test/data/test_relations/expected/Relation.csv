Expression	Identifier	Line	Column	Type
"(var6) <+ (var7)"	226	52	18	(POW (PRODUCT BOOL INTEGER))
"(var6) |>> ({1,2})"	222	51	18	(POW (PRODUCT BOOL INTEGER))
"{1,2}"	221	51	22	(POW INTEGER)
"1"	219	51	23	INTEGER
"(var5) |> (INTEGER)"	215	50	18	(POW (PRODUCT BOOL INTEGER))
"TRUE"	211	49	21	BOOL
"FALSE"	210	49	14	BOOL
"(BOOL) <<| (inv)"	209	48	18	(POW (PRODUCT BOOL INTEGER))
"{TRUE,FALSE}"	104	27	9	(POW BOOL)
"({0,1}) prj1 ({FALSE,TRUE})"	170	40	13	(POW (PRODUCT (PRODUCT INTEGER BOOL) INTEGER))
"1"	98	26	18	INTEGER
"ran(var3)"	97	26	5	(POW INTEGER)
"TRUE"	94	25	18	BOOL
"proj2"	5	9	33	(POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL))
"(INTEGER) *s (BOOL)"	40	18	24	(POW (PRODUCT INTEGER BOOL))
"POW((INTEGER) *s (INTEGER))"	89	24	19	(POW (POW (PRODUCT INTEGER INTEGER)))
"dom(var3)"	91	25	5	(POW BOOL)
"var2"	1	15	5	(POW (PRODUCT INTEGER INTEGER))
"POW((INTEGER) *s (INTEGER))"	84	23	19	(POW (POW (PRODUCT INTEGER INTEGER)))
"POW(((BOOL) *s (REAL)) *s ((INTEGER) *s (BOOL)))"	74	21	18	(POW (POW (PRODUCT (PRODUCT BOOL REAL) (PRODUCT INTEGER BOOL))))
"((BOOL) *s (REAL)) *s ((INTEGER) *s (BOOL))"	71	21	36	(POW (PRODUCT (PRODUCT BOOL REAL) (PRODUCT INTEGER BOOL)))
"FALSE"	103	27	16	BOOL
"{TRUE,FALSE}"	96	25	17	(POW BOOL)
"(INTEGER) *s (INTEGER)"	86	24	31	(POW (PRODUCT INTEGER INTEGER))
"(BOOL) *s (REAL)"	65	21	28	(POW (PRODUCT BOOL REAL))
"(INTEGER) *s (REAL)"	57	20	39	(POW (PRODUCT INTEGER REAL))
"(BOOL) *s ((INTEGER) *s (REAL))"	60	20	28	(POW (PRODUCT BOOL (PRODUCT INTEGER REAL)))
"TRUE"	102	27	10	BOOL
"{FALSE,TRUE}"	212	49	13	(POW BOOL)
"{1,2,3}"	101	26	17	(POW INTEGER)
"REAL"	56	20	41	(POW REAL)
"(INTEGER) *s (INTEGER)"	76	22	24	(POW (PRODUCT INTEGER INTEGER))
"POW((BOOL) *s (INTEGER))"	135	32	12	(POW (POW (PRODUCT BOOL INTEGER)))
"(INTEGER) *s (BOOL)"	48	19	25	(POW (PRODUCT INTEGER BOOL))
"POW(((INTEGER) *s (BOOL)) *s (INTEGER))"	46	18	12	(POW (POW (PRODUCT (PRODUCT INTEGER BOOL) INTEGER)))
"rfl_closure"	9	10	36	(POW (PRODUCT INTEGER INTEGER))
"TRUE"	168	40	33	BOOL
"closure(var2)"	207	46	20	(POW (PRODUCT INTEGER INTEGER))
"paral_prod"	7	10	18	(POW (PRODUCT (PRODUCT BOOL REAL) (PRODUCT INTEGER BOOL)))
"2"	99	26	20	INTEGER
"POW(POW((INTEGER) *s (BOOL)))"	23	14	11	(POW (POW (POW (PRODUCT INTEGER BOOL))))
"POW((BOOL) *s (INTEGER))"	125	30	12	(POW (POW (PRODUCT BOOL INTEGER)))
"inv"	3	9	22	(POW (PRODUCT BOOL INTEGER))
"POW((BOOL) *s (INTEGER))"	130	31	12	(POW (POW (PRODUCT BOOL INTEGER)))
"(INTEGER) *s (INTEGER)"	25	15	24	(POW (PRODUCT INTEGER INTEGER))
"FALSE"	176	41	27	BOOL
"closure1(var2)"	208	47	20	(POW (PRODUCT INTEGER INTEGER))
"POW((BOOL) *s (INTEGER))"	33	16	12	(POW (POW (PRODUCT BOOL INTEGER)))
"iter"	8	10	30	(POW (PRODUCT INTEGER INTEGER))
"(BOOL) *s (INTEGER)"	117	29	21	(POW (PRODUCT BOOL INTEGER))
"{(FALSE) |-> (0.0),(TRUE) |-> (1.0)}"	191	43	28	(POW (PRODUCT BOOL REAL))
"FALSE"	167	40	26	BOOL
"var3"	2	9	16	(POW (PRODUCT BOOL INTEGER))
"POW((BOOL) *s (INTEGER))"	120	29	12	(POW (POW (PRODUCT BOOL INTEGER)))
"(INTEGER) *s (BOOL)"	68	21	47	(POW (PRODUCT INTEGER BOOL))
"var4"	11	11	18	(POW (PRODUCT BOOL INTEGER))
"POW((INTEGER) *s (INTEGER))"	79	22	12	(POW (POW (PRODUCT INTEGER INTEGER)))
"(inv) [ ({TRUE,FALSE})"	105	27	5	(POW INTEGER)
"direct_prod"	6	10	5	(POW (PRODUCT BOOL (PRODUCT INTEGER REAL)))
"POW((BOOL) *s ((INTEGER) *s (REAL)))"	63	20	19	(POW (POW (PRODUCT BOOL (PRODUCT INTEGER REAL))))
"(BOOL) *s (INTEGER)"	35	17	20	(POW (PRODUCT BOOL INTEGER))
"var"	0	9	5	(POW (POW (PRODUCT INTEGER BOOL)))
"var6"	13	11	30	(POW (PRODUCT BOOL INTEGER))
"FALSE"	95	25	24	BOOL
"var7"	14	11	36	(POW (PRODUCT BOOL INTEGER))
"var5"	12	11	24	(POW (PRODUCT BOOL INTEGER))
"POW((BOOL) *s (INTEGER))"	38	17	11	(POW (POW (PRODUCT BOOL INTEGER)))
"1"	174	41	22	INTEGER
"INTEGER"	16	14	19	(POW INTEGER)
"BOOL"	17	14	29	(POW BOOL)
"(INTEGER) *s (BOOL)"	18	14	27	(POW (PRODUCT INTEGER BOOL))
"2"	220	51	25	INTEGER
"closure_var"	10	11	5	(POW (PRODUCT INTEGER INTEGER))
"POW((INTEGER) *s (INTEGER))"	28	15	12	(POW (POW (PRODUCT INTEGER INTEGER)))
"((INTEGER) *s (BOOL)) *s (BOOL)"	51	19	32	(POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL))
"(BOOL) *s (INTEGER)"	30	16	21	(POW (PRODUCT BOOL INTEGER))
"POW((INTEGER) *s (BOOL))"	21	14	15	(POW (POW (PRODUCT INTEGER BOOL)))
"1"	109	27	27	INTEGER
"2"	110	27	30	INTEGER
"{1,2}"	111	27	26	(POW INTEGER)
"0.0"	195	44	30	REAL
"POW((BOOL) *s (INTEGER))"	115	28	12	(POW (POW (PRODUCT BOOL INTEGER)))
"(BOOL) *s (INTEGER)"	122	30	21	(POW (PRODUCT BOOL INTEGER))
"proj"	4	9	27	(POW (PRODUCT (PRODUCT INTEGER BOOL) INTEGER))
"POW((BOOL) *s (INTEGER))"	140	33	12	(POW (POW (PRODUCT BOOL INTEGER)))
"0"	142	37	12	INTEGER
"5"	143	37	15	INTEGER
"(BOOL) *s (INTEGER)"	137	33	21	(POW (PRODUCT BOOL INTEGER))
"(0) .. (5)"	144	37	13	(POW INTEGER)
"{(0.0) |-> (FALSE),(1.0) |-> (TRUE)}"	201	44	28	(POW (PRODUCT REAL BOOL))
"3"	100	26	22	INTEGER
"(inv) ; (var2)"	182	42	18	(POW (PRODUCT BOOL INTEGER))
"1"	148	38	17	INTEGER
"2"	149	38	19	INTEGER
"TRUE"	188	43	47	BOOL
"({FALSE,TRUE}) <| (inv)"	213	49	27	(POW (PRODUCT BOOL INTEGER))
"(BOOL) *s (INTEGER)"	132	32	21	(POW (PRODUCT BOOL INTEGER))
"3"	150	38	21	INTEGER
"(var2) iterate (2)"	206	45	13	(POW (PRODUCT INTEGER INTEGER))
"((INTEGER) *s (BOOL)) *s (INTEGER)"	43	18	31	(POW (PRODUCT (PRODUCT INTEGER BOOL) INTEGER))
"{1,2,3}"	151	38	16	(POW INTEGER)
"(1) |-> (TRUE)"	159	39	31	(PRODUCT INTEGER BOOL)
"id({1,2,3})"	152	38	13	(POW (PRODUCT INTEGER INTEGER))
"TRUE"	177	41	34	BOOL
"0"	154	39	14	INTEGER
"(INTEGER) *s (INTEGER)"	81	23	31	(POW (PRODUCT INTEGER INTEGER))
"FALSE"	155	39	20	BOOL
"(0) |-> (FALSE)"	156	39	16	(PRODUCT INTEGER BOOL)
"1"	157	39	29	INTEGER
"1"	165	40	21	INTEGER
"(BOOL) *s (INTEGER)"	112	28	21	(POW (PRODUCT BOOL INTEGER))
"TRUE"	158	39	35	BOOL
"{0,1}"	166	40	18	(POW INTEGER)
"{(0) |-> (FALSE),(1) |-> (TRUE)}"	160	39	12	(POW (PRODUCT INTEGER BOOL))
"POW(((INTEGER) *s (BOOL)) *s (BOOL))"	54	19	13	(POW (POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL)))
"~({(0) |-> (FALSE),(1) |-> (TRUE)})"	161	39	12	(POW (PRODUCT BOOL INTEGER))
"0"	164	40	19	INTEGER
"{FALSE,TRUE}"	169	40	25	(POW BOOL)
"{0,1}"	175	41	19	(POW INTEGER)
"(BOOL) *s (INTEGER)"	127	31	21	(POW (PRODUCT BOOL INTEGER))
"{FALSE,TRUE}"	178	41	26	(POW BOOL)
"TRUE"	199	44	55	BOOL
"1.0"	198	44	47	REAL
"({0,1}) prj2 ({FALSE,TRUE})"	179	41	14	(POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL))
"2"	205	45	27	INTEGER
"FALSE"	185	43	30	BOOL
"0.0"	186	43	40	REAL
"1.0"	189	43	56	REAL
"(TRUE) |-> (1.0)"	190	43	52	(PRODUCT BOOL REAL)
"(FALSE) |-> (0.0)"	187	43	36	(PRODUCT BOOL REAL)
"(var3) >< ({(FALSE) |-> (0.0),(TRUE) |-> (1.0)})"	192	43	25	(POW (PRODUCT BOOL (PRODUCT INTEGER REAL)))
"FALSE"	196	44	38	BOOL
"0"	173	41	20	INTEGER
"(var3) || ({(0.0) |-> (FALSE),(1.0) |-> (TRUE)})"	202	44	25	(POW (PRODUCT (PRODUCT BOOL REAL) (PRODUCT INTEGER BOOL)))
"(0.0) |-> (FALSE)"	197	44	34	(PRODUCT REAL BOOL)
"var8"	15	11	42	(POW (PRODUCT BOOL INTEGER))
"((0) .. (5)) <-> (BOOL)"	145	37	17	(POW (POW (PRODUCT INTEGER BOOL)))
"(1.0) |-> (TRUE)"	200	44	51	(PRODUCT REAL BOOL)
